Add compile and platform yamls to compile model with FRE/2025.01

uwagura opened the pull request:
This PR adds compile and platform yamls to allow users to compile the NWA configuration with `Fre/2025.01`, and optionally produce containers to run the model as well. Both the containers and the bare metal models compile with these yamls can then be run using `fre/bronx-23`. 

The full containerized does not quite reproduce the bare-metal run results, but the differences mostly seem minimal: 


![sst_eval_nwa_container](https://github.com/user-attachments/assets/67f71421-7f8d-410a-b1a3-f87d6b3bcf24)
![mean_ssh_eval_nwa_container](https://github.com/user-attachments/assets/a4ec3041-eb3c-4f9e-b327-ad15bdaa1dd6)
![sst_trends_nwa_container](https://github.com/user-attachments/assets/3993c2eb-488d-4499-a1c6-59227ffe05e7)
![domain_chl_nwa_container](https://github.com/user-attachments/assets/c785ae26-4a28-4b51-976a-a550b2c9c66a)
![regional_chl_nwa_container](https://github.com/user-attachments/assets/b3eb1e86-f0a1-49d4-b27e-e8d5d6298bd9)

yichengt900 commented:
@uwagura, thanks for opening this PR—awesome work! Although it is still a draft, I have two quick thoughts:

Would you mind adding a README file under yaml with brief instructions on how to perform bare-metal/container builds and how to run containerized builds using `bronx-23`?

When you get a chance, could you also share the archive paths for both your bare-metal and containerized NWA-12 runs?
Thanks!

uwagura commented:
@yichengt900 , I've added a README to the yaml directory, and added instructions to the main yaml on how to use `fre make`. Let me know what you think.

You can find the history files and the pp files from this experiment here: `/archive/Utheri.Wagura/fre/cefi/NWA/FullContainerRun/CEFI_NWA12_COBALT_V1/gfdl.ncrc5-intel23-prod/`

@andrew-c-ross , you don't have to try the containerized run if you don't have the time, but I'd love to get your thoughts on some of the plots!

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
+# COMPILE GUIDE - BAREMETAL RUN
+# 
+# 1.) Create script that will be used to checkout model components
+# fre make create-checkout -y CEFI_NWA12_cobalt.yaml -p hpcme.2023 -t prod
```
`hpcme.2023` should be `ncrc5.intel23` or `ncrc6.intel23`

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
+# fre make create-checkout -y CEFI_NWA12_cobalt.yaml -p hpcme.2023 -t prod
+# 
+# 2.) Create Makefile
+# fre make create-makefile -y CEFI_NWA12_cobalt.yaml -p hpcme.2023 -t prod
```
Same as above.

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
-# Quick start instructions and source code can be found here: https://github.com/NOAA-GFDL/fre-cli/tree/main/fre/pp
+# 3.) Create script that will compile the model, and OPTIONALLY run it with the --execute flag. You can manually run the 
+# compile script after this step as well
+# fre make create-compile -y CEFI_NWA12_cobalt.yaml -p hpcme.2023 -t prod --execute
```
Same.

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
 #
-# For modularity and compatibility with with cylc workflow engine, fre/canopy splits up the post processcing process into several sub steps:
+# 1.) Create checkout script with the "no parallel checkouts option" - this is required for container builds
+# fre make create-checkout -y CEFI_NWA12_cobalt.yaml -p hpcme.2023 -t prod --npc
```
should be `-npc`

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
 # fre pp checkout -e CEFI_NWA12_COBALT_V1 -p gfdl.ncrc5-intel22 -t prod
 #
 # 2.) Combine your main yaml and experiment yamls into a single yaml, then set up the cylc-src dir with the configure-yaml command:
 # fre yamltools combine-yamls -e CEFI_NWA12_COBALT_V1 -p gfdl.ncrc5-intel22 -t prod -y CEFI_NWA12_cobalt.yaml
```
Need `--use pp` flag

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -1,22 +1,47 @@
 #
 # 2.) Combine your main yaml and experiment yamls into a single yaml, then set up the cylc-src dir with the configure-yaml command:
 # fre yamltools combine-yamls -e CEFI_NWA12_COBALT_V1 -p gfdl.ncrc5-intel22 -t prod -y CEFI_NWA12_cobalt.yaml
 # fre pp configure-yaml -e CEFI_NWA12_COBALT_V1 -p gfdl.ncrc5-intel22 -t prod -y combined-CEFI_NWA12_COBALT_V1.yaml
```
This did not work on my end. I have to change it to fre pp configure-yaml -e CEFI_NWA12_COBALT_V1 -p gfdl.ncrc5-intel22 -t prod -y CEFI_NWA12_cobalt.yaml`

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -47,7 +71,17 @@ fre_properties:
+    - &momIncludes    "-Imom6/src/MOM6/pkg/CVMix-src/include"
+    - &sisincludes    "-Imom6/src/MOM6/src/framework"
+    - &MOM6_GIT_FIX   "main"
+    - &FMS_GIT_TAG    "2024.01.02"
```
Please update this to `2024.03`, which is the one we used in xml: https://github.com/NOAA-GFDL/CEFI-regional-MOM6/blob/d832842dda40dd743a0df7c669edaa10163619bf/xmls/NWA12/CEFI_NWA12_cobalt.xml#L43

yichengt900 commented on yamls/NWA12/CEFI_NWA12_cobalt.yaml:
```diff
@@ -47,7 +71,17 @@ fre_properties:
+    - &sisincludes    "-Imom6/src/MOM6/src/framework"
+    - &MOM6_GIT_FIX   "main"
+    - &FMS_GIT_TAG    "2024.01.02"
+    - &CPL_GIT_TAG    "2024.01.02"
```
Also update this one to `2024.03`

yichengt900 commented on yamls/NWA12/README.md:
````diff
@@ -0,0 +1,112 @@
+```
+
+
+## Postprocessing the Experiment
````
Please update the `postprocess section` by following my comments above.

yichengt900 commented on yamls/NWA12/platforms.yaml:
```diff
@@ -0,0 +1,49 @@
+platforms:
+   - name: ncrc5.intel22
```
Since now the default Intel compiler is `2023.02`, we can consider remove ncrc*.intel22.

yichengt900 commented on yamls/NWA12/platforms.yaml:
```diff
@@ -0,0 +1,49 @@
+   - name: ncrc5.intel23
+     compiler: intel
+     modulesInit: [" module use -a /ncrc/home2/fms/local/modulefiles \n","source $MODULESHOME/init/sh \n"]
+     modules: [ !join [*INTEL, "/2023.1.0"],"fre/bronx-22",cray-hdf5, cray-netcdf]
```
The new default should be `2023.2.0`.  Also I think we do not need `fre/bronx-22` here?

uwagura replied:
It looks like we still need some version of `fre` here - when `fre make create-compile` makes the compile script, it uses `module loads` the modules listed here, so it needs access to some version of fre to access `list_paths` and `mkmf`

yichengt900 commented on yamls/NWA12/platforms.yaml:
```diff
@@ -0,0 +1,49 @@
+     #fc: ftn
+     #cc: cc
+     mkTemplate: !join ["/ncrc/home2/fms/local/opt/fre-commands/bronx-22/site/ncrc5/", *INTEL, ".mk"]
+     modelRoot: /gpfs/f5/cefi/scratch/Utheri.Wagura/fre/cefi/NWA/Canopy_202412/
```
May consider change this to `!join ["/gpfs/f5/cefi/scratch/${USER}", "/", *FRE_STEM]`. Same for the `ncrc6.intel23` platform

yichengt900 reviewed (changes requested):
@uwagura, thank you for opening this PR. Overall, it looks good, and I only have a few comments regarding your instructions and some other minor things.

I successfully built the containerized model by following the modified instructions with `fre/2025.01`. Additionally, I conducted several 2-day regression test runs using both the bare-metal build and the containerized build, and the restarts were identical.

Furthermore, I compared the model results from your fully containerized run with my `2024/12` retrospective runs. Based on my analysis, the model results also appear to be identical. Relevant figures can be found here:

**2024/12 bare-metal runs**: `/nbhome/ynt/results/CEFI_NWA12_COBALT_V1_repro/analysis/regional-mom6-diagnostics_20240418/physics/figures_2024_12`

**Fully containerized runs**: `/nbhome/ynt/results/CEFI_NWA12_COBALT_V1_repro/analysis/regional-mom6-diagnostics_20240418/physics/figures_container`

Given these findings, I’m unsure what you meant by your original comment in your PR: `The fully containerized run does not quite reproduce the bare-metal run results, but the differences mostly seem minimal.` Are you referring to differences compared to the original NWA paper results?

uwagura commented:
Thanks for your review @yichengt900 , I'll make the changes you suggested soon. 

Regarding my comment, about the results not being quite the same, I meant that the results from the containerized run were not quite identical to the plots I produced with the data in `/archive/acr/fre/NWA/2023_04/NWA12_COBALT_2023_04_kpo4-coastatten-physics/gfdl.ncrc5-intel22-prod`. For example, compare the above `sst` plot with the plot in #74 , or the`ssh` plot with the one in #101. None of the differences seem particularly big or concerning, just thought that I'd mention them.

yichengt900 commented:
> Thanks for your review @yichengt900 , I'll make the changes you suggested soon.
> 
> Regarding my comment, about the results not being quite the same, I meant that the results from the containerized run were not quite identical to the plots I produced with the data in `/archive/acr/fre/NWA/2023_04/NWA12_COBALT_2023_04_kpo4-coastatten-physics/gfdl.ncrc5-intel22-prod`. For example, compare the above `sst` plot with the plot in #74 , or the`ssh` plot with the one in #101. None of the differences seem particularly big or concerning, just thought that I'd mention them.

@uwagura, thank you for the clarification. It seems the data you pointed to is likely the same one Andrew used for his paper. Unfortunately, it’s probably not possible to fully reproduce those results due to updates in libraries and compilers on C5, as well as changes in the model codes.

Our main goal here is to confirm that the containerized version produces the same results as the bare-metal build. One good thing is even with those differences, the new model results are consistent and not worse than the ones in the paper.

yichengt900 reviewed (commented):
@uwagura, everything looks good now except for a missing `. I will approve it after you change that. Thanks again for your contribution
